File: apps/api/app/expenses/application/create_expense.py
"""Use case: record a new expense."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date

from app.expenses.domain.entities import DraftExpense, Expense
from app.expenses.domain.errors import UnknownCategoryError
from app.expenses.ports.category_checker import CategoryChecker
from app.expenses.ports.repository import ExpenseRepository
from app.shared.domain.money import Money


@dataclass(frozen=True, slots=True)
class CreateExpenseCommand:
    amount_cents: int
    currency: str
    category_id: int
    occurred_on: date
    note: str | None = None


class CreateExpense:
    def __init__(self, repository: ExpenseRepository, categories: CategoryChecker) -> None:
        self._repository = repository
        self._categories = categories

    def __call__(self, command: CreateExpenseCommand) -> Expense:
        if not self._categories.exists(command.category_id):
            raise UnknownCategoryError(command.category_id)
        draft = DraftExpense(
            money=Money(command.amount_cents, command.currency),
            category_id=command.category_id,
            occurred_on=command.occurred_on,
            note=command.note,
        )
        return self._repository.add(draft)

File: apps/api/internal/expenses/application/create_expense.go
package application

import (
	"context"

	"caudal-api/internal/expenses/domain"
	"caudal-api/internal/expenses/ports"
	shared "caudal-api/internal/shared/domain"
)

type CreateExpense struct {
	repository ports.Repository
	checker    ports.CategoryChecker
}

func NewCreateExpense(repository ports.Repository, checker ports.CategoryChecker) CreateExpense {
	return CreateExpense{repository: repository, checker: checker}
}

func (useCase CreateExpense) Execute(ctx context.Context, command CreateExpenseCommand) (domain.Expense, error) {
	exists, err := useCase.checker.ExpenseCategoryExists(ctx, command.CategoryID)
	if err != nil {
		return domain.Expense{}, err
	}
	if !exists {
		return domain.Expense{}, domain.ErrUnknownCategory
	}
	money, err := shared.NewMoney(command.AmountCents, command.Currency)
	if err != nil {
		return domain.Expense{}, err
	}
	expense, err := domain.NewExpense(0, money, command.CategoryID, command.OccurredOn, command.Note)
	if err != nil {
		return domain.Expense{}, err
	}
	return useCase.repository.Create(ctx, expense)
}
